File: module-07-transport-layer/udp_echo_client.js
/**
 * Module 07 — JS/Node port of udp_echo_client.py. Sends 15 datagrams,
 * uses a timeout to detect replies that never arrive -- no retry logic,
 * exactly like the Python variant.
 */
const dgram = require("dgram");

const HOST = "127.0.0.1";
const PORT = 9019;
const TIMEOUT_MS = 500;

const socket = dgram.createSocket("udp4");
let sent = 0;
let received = 0;

function sendOne(i) {
  return new Promise((resolve) => {
    const message = Buffer.from(`datagram #${i}`);
    let settled = false;

    const timer = setTimeout(() => {
      if (!settled) {
        settled = true;
        console.log(`  #${i}: sent ${JSON.stringify(message.toString())} -> NO REPLY (lost, and nothing resends it for you)`);
        socket.removeListener("message", onMessage);
        resolve();
      }
    }, TIMEOUT_MS);

    function onMessage(reply) {
      if (!settled) {
        settled = true;
        clearTimeout(timer);
        received++;
        console.log(`  #${i}: sent ${JSON.stringify(message.toString())} -> got reply ${JSON.stringify(reply.toString())}`);
        resolve();
      }
    }

    socket.once("message", onMessage);
    socket.send(message, PORT, HOST);
    sent++;
  });
}

async function main() {
  for (let i = 1; i <= 15; i++) {
    await sendOne(i);
  }
  const lossPct = Math.round((100 * (sent - received)) / sent);
  console.log(`\nSent ${sent} datagrams, received ${received} replies (${sent - received} lost, ${lossPct}% loss).`);
  console.log("Compare this to tcp_echo_client.js, which got all 5 of its");
  console.log("messages back with zero code written to handle loss -- because");
  console.log("TCP's retransmission logic handled it below the application layer.");
  socket.close();
}

main();

File: module-07-transport-layer/udp_echo_client.py
"""
Module 07 -- UDP echo client. Sends 15 numbered datagrams and reports
which replies never arrive -- no retry, no reordering logic, nothing.
This IS what "unreliable, connectionless" means in practice, not just in
the textbook definition.
"""
import socket

HOST, PORT = "127.0.0.1", 9009


def main():
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.settimeout(0.5)  # if a reply doesn't arrive quickly, it's not coming

        sent, received = 0, 0
        for i in range(1, 16):
            message = f"datagram #{i}".encode()
            sock.sendto(message, (HOST, PORT))
            sent += 1
            try:
                reply, _ = sock.recvfrom(1024)
                received += 1
                print(f"  #{i}: sent {message!r} -> got reply {reply!r}")
            except socket.timeout:
                print(f"  #{i}: sent {message!r} -> NO REPLY (lost, and nothing resends it for you)")

        print(f"\nSent {sent} datagrams, received {received} replies "
              f"({sent - received} lost, {100 * (sent - received) / sent:.0f}% loss).")
        print("Compare this to tcp_echo_client.py, which got all 5 of its")
        print("messages back with zero code written to handle loss -- because")
        print("TCP's retransmission logic handled it below the application layer.")


if __name__ == "__main__":
    main()
